test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,258 @@
+    compute_arg_dict, decorate_test_fn, compute_temp_file_path, generate_error_msg, is_torch_nn_functional_test
+import cpp_api_parity
+
+# Expected substitutions:
+#
+# ${functional_variant_name}  (e.g. `BCELoss_no_reduce`)
+# ${cpp_args_construction_stmts}
+# ${cpp_function_call}
+TORCH_NN_FUNCTIONAL_TEST_FORWARD = Template("""
+void ${functional_variant_name}_test_forward(
+    const std::string& arg_dict_file_path,
+    const std::string& forward_output_file_path) {
+  pybind11::gil_scoped_release no_gil;
+
+  namespace F = torch::nn::functional;
+
+  // Declare arguments
+  auto arg_dict = load_dict_from_file(arg_dict_file_path);
+  ${cpp_args_construction_stmts};
+
+  // Some functionals (such as `F::rrelu`) create random tensors in their call path.
+  // To make sure the random tensors created are the same in Python/C++, we need
+  // to set the RNG seed manually.
+  torch::manual_seed(0);
+
+  // Run function with arguments
+  auto cpp_output = ${cpp_function_call};
+
+  // Save the output into a file to be compared in Python later
+  write_ivalue_to_file(torch::IValue(cpp_output), forward_output_file_path);
+}
+""")
+
+def run_forward(unit_test_class, test_params):
+    device = test_params.device
+
+    inputs = set_python_tensors_requires_grad(move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['input']], device))
+    inputs = inputs + move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['target']], device)
+    inputs = inputs + move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['extra_args']], device)
+
+    # Some functionals (such as `F.rrelu`) create random tensors in their call path.
+    # To make sure the random tensors created are the same in Python/C++, we need
+    # to set the RNG seed manually.
+    torch.manual_seed(0)
+    python_output = test_params.test_instance.constructor()(*inputs)
+
+    return python_output
+
+def test_forward(unit_test_class, test_params):
+    functional_variant_name = test_params.functional_variant_name
+    cpp_tmp_folder = test_params.cpp_tmp_folder
+
+    # Run forward on Python functional
+    python_output = run_forward(unit_test_class, test_params)
+
+    # Save Python arguments to be used from C++ function
+    arg_dict_file_path = compute_temp_file_path(cpp_tmp_folder, functional_variant_name, 'arg_dict')
+    serialize_arg_dict_as_script_module(test_params.arg_dict).save(arg_dict_file_path)
+
+    cpp_test_name = '{}_{}'.format(test_params.functional_variant_name, 'test_forward')
+    cpp_test_fn = getattr(unit_test_class.functional_impl_check_cpp_module, cpp_test_name)
+
+    def run_cpp_test_fn_and_check_output():
+        forward_output_file_path = compute_temp_file_path(cpp_tmp_folder, functional_variant_name, 'forward_output')
+
+        cpp_test_fn(arg_dict_file_path, forward_output_file_path)
+        cpp_output = torch.load(forward_output_file_path)
```

Comment:
nit: There's a lot of I/O going on with these tests, which increases the risk of flakiness. If that becomes a problem in the future maybe someone can use pybind to connect the result of the cpp api to python

Reply:
Sounds great and I noted it in the follow-up items to do